File: frontend/src/lib/wellbeing.ts
/**
 * Wellbeing Profile: six capacities on one 0-100 scale, higher is always better.
 * Mirrors backend/wellbeing.py so a single entry's profile can be drawn straight from
 * the feedback payload, while period averages come from GET /api/insights.
 */

export interface WellbeingAxis {
  key: string
  label: string
  description: string
  value: number | null
  previous?: number | null
}

/** Plain-language background for each axis: what it is, why it matters, how this app measures it, where to read more. */
export interface AxisGuide {
  what: string
  why: string
  measured: string
  moves: string
  link: { label: string; href: string }
}

export const WELLBEING_AXES: Omit<WellbeingAxis, "value" | "previous">[] = [
  { key: "mood", label: "Mood", description: "Emotional state rated by the analysis, 1 to 10." },
  { key: "energy", label: "Energy", description: "The Find Your Energy battery level." },
  { key: "calm", label: "Calm", description: "The inverse of the rumination level: little overthinking scores high." },
  { key: "agency", label: "Agency", description: "Share of what you wrote about that sits within your control." },
  { key: "outward", label: "Outward focus", description: "The inverse of the self-focus score: attention on the world, not only on yourself." },
  { key: "clarity", label: "Clarity", description: "Linguistic clarity, the grammar score." },
]

export const AXIS_GUIDES: Record<string, AxisGuide> = {
  mood: {
    what: "How you were feeling, as read from the tone of the entry, on a scale from despair to euphoric.",
    why: "Mood colours attention, memory and judgement. A low mood makes problems look bigger and options fewer, so knowing where you are helps you discount the distortion.",
    measured: "The analysis rates each entry from 1 to 10; this axis is that rating as a percentage, averaged over the period.",
    moves: "Sleep, movement, daylight, contact with people you like, and finishing things you have been avoiding.",
    link: { label: "Mood (psychology), Wikipedia", href: "https://en.wikipedia.org/wiki/Mood_(psychology)" },
  },
  energy: {
    what: "Your mental battery: how much capacity the day left you with.",
    why: "Energy is the budget everything else is paid from. When it is low, small tasks cost more and self-control runs out sooner, so protecting what charges you is the highest-leverage habit there is.",
    measured: "The Find Your Energy battery: mood, the balance of chargers to drainers, how much of the entry is within your control, focus, clarity and engagement, with a bonus for completed micro-actions and resets.",
    moves: "Rest, cutting a known drainer, and doing one charging activity on purpose rather than by accident.",
    link: { label: "Vitality, Wikipedia", href: "https://en.wikipedia.org/wiki/Vitality" },
  },
  calm: {
    what: "How little your mind was looping: replaying the past or rehearsing the future without getting anywhere.",
    why: "Rumination is one of the strongest predictors of anxiety and low mood, and it feeds on itself. The skill is not stopping thoughts but noticing the loop early and stepping out of it, which is what the 3-Minute Reset practises.",
    measured: "The analysis rates rumination low, moderate or high for each entry; low scores 100, moderate 50, high 0.",
    moves: "The reset, a walk, writing the loop down once and closing it, and sleep.",
    link: { label: "Rumination (psychology), Wikipedia", href: "https://en.wikipedia.org/wiki/Rumination_(psychology)" },
  },
  agency: {
    what: "The share of what you wrote about that sits within your control, as opposed to things you can only worry about.",
    why: "People who believe their actions matter, an internal locus of control, report less stress and follow through more. The Stoic version is older: spend effort only where it can change something. Low agency days are the ones where most of the entry is about other people's choices.",
    measured: "From the Circle of Control: the number of controllable factors the analysis found, divided by all factors it found.",
    moves: "Restating a worry as the one action you could take, and letting the rest be outside your hands.",
    link: { label: "Locus of control, Wikipedia", href: "https://en.wikipedia.org/wiki/Locus_of_control" },
  },
  outward: {
    what: "How much of your attention was on the world and other people rather than only on yourself.",
    why: "Heavy self-focus travels with worry and low mood; attention that reaches outward, to people, work and surroundings, tends to lift wellbeing. It is not about ignoring yourself, only about not being trapped in the mirror.",
    measured: "The inverse of the self-focus score: an entry rated 3 out of 10 for self-focus scores 70 here.",
    moves: "Writing about someone else's day, gratitude, and doing something for another person.",
    link: { label: "Self-consciousness, Wikipedia", href: "https://en.wikipedia.org/wiki/Self-consciousness" },
  },
  clarity: {
    what: "How clearly the entry reads: sentences that hold together, grammar that does not trip.",
    why: "Clear writing and clear thinking move together. When a tangled mind produces tangled sentences, the fixes the analysis offers are a way back to the thought. Expressive writing research links putting experiences into coherent words with better health over time.",
    measured: "The grammar score from the analysis, 1 to 10, as a percentage.",
    moves: "Shorter sentences, one idea each, and reading the entry back once.",
    link: { label: "Expressive writing, Wikipedia", href: "https://en.wikipedia.org/wiki/Expressive_writing" },
  },
}

const RUMINATION_TO_CALM: Record<string, number> = { low: 100, moderate: 50, high: 0 }

const clamp = (v: number) => Math.max(0, Math.min(100, v))
const pct10 = (v: unknown) => (typeof v === "number" ? clamp((v / 10) * 100) : null)
const round1 = (v: number) => Math.round(v * 10) / 10

/** Profile of one analysed entry, from the feedback object the analyze endpoint returns. */
export function profileFromFeedback(feedback: any): WellbeingAxis[] {
  const energy = feedback?.energyData || {}
  const controllables = Array.isArray(energy.controllables) ? energy.controllables.length : 0
  const uncontrollables = Array.isArray(energy.uncontrollables) ? energy.uncontrollables.length : 0
  const total = controllables + uncontrollables
  const rumination = typeof energy.rumination_level === "string" ? energy.rumination_level.toLowerCase() : null

  const values: Record<string, number | null> = {
    mood: pct10(feedback?.moodScore),
    energy: typeof energy.battery_level === "number" ? clamp(energy.battery_level) : null,
    calm: rumination !== null && rumination in RUMINATION_TO_CALM ? RUMINATION_TO_CALM[rumination] : null,
    agency: total > 0 ? (controllables / total) * 100 : null,
    outward: typeof feedback?.selfFocusScore === "number" ? clamp(((10 - feedback.selfFocusScore) / 10) * 100) : null,
    clarity: pct10(feedback?.grammarScore),
  }

  return WELLBEING_AXES.map(axis => {
    const v = values[axis.key]
    return { ...axis, value: v === null || v === undefined ? null : round1(v) }
  })
}

/** Enough signals to be worth drawing (a radar with one or two points is misleading). */
export function hasProfile(axes: WellbeingAxis[]): boolean {
  return axes.filter(a => a.value !== null).length >= 3
}

File: backend/wellbeing.py
"""
wellbeing.py - the Wellbeing Profile: six capacities on one 0-100 scale, higher is always better.

Every analysed entry already carries the raw signals (mood, grammar, self-focus, battery,
rumination, controllables). This module turns them into a fixed set of axes so they can be
drawn as a radar and compared across periods. The axis list and the mapping are mirrored in
frontend/src/lib/wellbeing.ts, which computes the same profile for a single entry client-side.

Axes point the same way on purpose: a bigger shape is always better, which is the one rule
a radar chart needs to be readable.
"""

from typing import Optional

WELLBEING_AXES = [
    {"key": "mood", "label": "Mood", "description": "Emotional state rated by the analysis, 1 to 10."},
    {"key": "energy", "label": "Energy", "description": "The Find Your Energy battery level."},
    {"key": "calm", "label": "Calm", "description": "The inverse of the rumination level: little overthinking scores high."},
    {"key": "agency", "label": "Agency", "description": "Share of what you wrote about that sits within your control."},
    {"key": "outward", "label": "Outward focus", "description": "The inverse of the self-focus score: attention on the world, not only on yourself."},
    {"key": "clarity", "label": "Clarity", "description": "Linguistic clarity, the grammar score."},
]

RUMINATION_TO_CALM = {"low": 100.0, "moderate": 50.0, "high": 0.0}


def _pct10(score) -> Optional[float]:
    if score is None:
        return None
    return max(0.0, min(100.0, float(score) / 10.0 * 100.0))


def axes_from_feedback(fb) -> dict:
    """Axis values for one FeedbackReport, or None where the analysis produced no signal."""
    energy = (getattr(fb, "energy_data", None) or {}) if fb is not None else {}
    controllables = energy.get("controllables") or []
    uncontrollables = energy.get("uncontrollables") or []
    n_control = len(controllables) + len(uncontrollables)

    rumination = energy.get("rumination_level")
    calm = RUMINATION_TO_CALM.get(rumination.lower()) if isinstance(rumination, str) else None

    battery = energy.get("battery_level")
    energy_value = max(0.0, min(100.0, float(battery))) if isinstance(battery, (int, float)) else None

    self_focus = getattr(fb, "self_focus_score", None) if fb is not None else None
    outward = max(0.0, min(100.0, (10.0 - float(self_focus)) / 10.0 * 100.0)) if self_focus is not None else None

    return {
        "mood": _pct10(getattr(fb, "mood_score", None) if fb is not None else None),
        "energy": energy_value,
        "calm": calm,
        "agency": (len(controllables) / n_control * 100.0) if n_control else None,
        "outward": outward,
        "clarity": _pct10(getattr(fb, "grammar_score", None) if fb is not None else None),
    }


def average_axes(feedbacks) -> tuple[dict, int]:
    """
    Average each axis over the entries that have it. Returns (values, entries), where values maps
    axis key to a rounded 0-100 number or None, and entries counts reports that contributed at all.
    """
    totals: dict = {a["key"]: [0.0, 0] for a in WELLBEING_AXES}
    contributing = 0
    for fb in feedbacks:
        axes = axes_from_feedback(fb)
        if not any(v is not None for v in axes.values()):
            continue
        contributing += 1
        for key, value in axes.items():
            if value is not None:
                totals[key][0] += value
                totals[key][1] += 1
    values = {key: (round(total / count, 1) if count else None) for key, (total, count) in totals.items()}
    return values, contributing
